File: common/src/main/java/dev/obscuria/fragmentum/registry/DeferredItem.java
package dev.obscuria.fragmentum.registry;

import net.minecraft.core.Holder;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.ItemLike;

import java.util.function.Supplier;

@SuppressWarnings("unused")
public class DeferredItem<T extends Item> extends Deferred<Item, T> implements ItemLike
{
    public DeferredItem(Supplier<Holder<Item>> supplier)
    {
        super(supplier);
    }

    @Override
    public Item asItem()
    {
        return get();
    }

    public ItemStack instantiate()
    {
        return asItem().getDefaultInstance();
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/api/registry/DeferredItem.kt
package dev.obscuria.fragmentum.api.registry

import net.minecraft.core.Holder
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.ItemLike
import java.util.function.Supplier

class DeferredItem<T : Item>(supplier: Supplier<Holder<Item>>)
    : Deferred<Item, T>(supplier), ItemLike
{
    override fun asItem(): T = value()

    fun instance(): ItemStack = value().defaultInstance
}